File: src/main/java/net/greenjab/fixedminecraft/registry/item/EchoFruitItem.java
package net.greenjab.fixedminecraft.registry.item;

import java.util.Optional;
import java.util.Set;
import net.minecraft.core.GlobalPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.gameevent.GameEvent;
import net.minecraft.world.phys.Vec3;
import org.jspecify.annotations.NonNull;

public class EchoFruitItem extends Item {
    public EchoFruitItem(Properties settings) {
        super(settings);
    }

    @Override
    public @NonNull ItemStack finishUsingItem(@NonNull ItemStack stack, @NonNull Level world, @NonNull LivingEntity user) {
        super.finishUsingItem(stack, world, user);
        if (user instanceof ServerPlayer serverPlayerEntity) {
            Optional<GlobalPos> deathOpt = serverPlayerEntity.getLastDeathLocation();
            if (!world.isClientSide() && deathOpt.isPresent()) {
                GlobalPos deathPos = deathOpt.get();
                double d = deathPos.pos().getX();
                double e = deathPos.pos().getY();
                double f = deathPos.pos().getZ();
                if (serverPlayerEntity.isPassenger()) {
                    serverPlayerEntity.stopRiding();
                }

                Vec3 vec3d = serverPlayerEntity.position();
                ServerLevel serverWorld = serverPlayerEntity.level().getServer().getLevel(serverPlayerEntity.getLastDeathLocation().get().dimension());
                if (serverWorld!=null) {
                    if (serverPlayerEntity.teleportTo(serverWorld, d, e, f, Set.of(), serverPlayerEntity.getYRot(), serverPlayerEntity.getXRot(), true)) {
                        while (!serverWorld.noCollision(serverPlayerEntity) &&
                               serverPlayerEntity.getY() < serverWorld.getMaxY()) {
                            serverPlayerEntity.setPos(serverPlayerEntity.getX(),
                                    serverPlayerEntity.getY() + 1.0, serverPlayerEntity.getZ());
                        }
                        world.gameEvent(GameEvent.TELEPORT, vec3d, GameEvent.Context.of(serverPlayerEntity));
                        SoundEvent soundEvent = SoundEvents.CHORUS_FRUIT_TELEPORT;
                        SoundSource soundCategory = SoundSource.PLAYERS;

                        world.playSound(null, vec3d.x(), vec3d.y(), vec3d.z(), soundEvent, soundCategory);
                        user.resetFallDistance();
                    }
                }
            }
        }
        return stack;
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/item/EchoFruitItem.kt
package net.greenjab.fixedminecraft.registry.item

import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvent
import net.minecraft.sound.SoundEvents
import net.minecraft.world.World
import net.minecraft.world.event.GameEvent

class EchoFruitItem(settings: Settings?) : Item(settings) {
    override fun finishUsing(stack: ItemStack, world: World, user: LivingEntity): ItemStack {
        val itemStack = super.finishUsing(stack, world, user)
        if (user is PlayerEntity) {
            val deathOpt = user.lastDeathPos
            if (!world.isClient && deathOpt.isPresent) {
                val deathPos = user.lastDeathPos.get()
                val d = deathPos.pos.x.toDouble()
                val e = deathPos.pos.y.toDouble()
                val f = deathPos.pos.z.toDouble()
                if (user.hasVehicle()) {
                    user.stopRiding()
                }

                val vec3d = user.getPos()
                val serverWorld: ServerWorld? = user.server!!.getWorld(user.lastDeathPos.get().dimension)
                if (user.teleport(serverWorld as ServerWorld, d, e, f, setOf(), user.getYaw(), user.getPitch())) {
                    while (!serverWorld.isSpaceEmpty(user) && user.getY() < serverWorld.topY.toDouble()) {
                        user.setPosition(user.getX(), user.getY() + 1.0, user.getZ())
                    }
                    world.emitGameEvent(GameEvent.TELEPORT, vec3d, GameEvent.Emitter.of(user as Entity))
                    val soundEvent: SoundEvent = SoundEvents.ITEM_CHORUS_FRUIT_TELEPORT
                    val soundCategory: SoundCategory = SoundCategory.PLAYERS

                    world.playSound(null as PlayerEntity?, vec3d.getX(), vec3d.getY(), vec3d.getZ(), soundEvent, soundCategory)
                    user.onLanding()

                    user.itemCooldownManager[this] = 20
                }
            }
        }

        return itemStack
    }
}
